contrib/checkpassword-0.90/checkpassword.8
.TH checkpassword 8
.SH NAME
checkpassword \- check a password
.SH SYNOPSIS
.B checkpassword
.I subprogram
[
.I args ...
]
.SH DESCRIPTION
.B checkpassword
reads descriptor 3 through end of file
and then closes descriptor 3.
There must be at most 512 bytes of data before end of file.

The information supplied on descriptor 3
is a login name terminated by \e0,
a password terminated by \e0,
a timestamp terminated by \e0,
and possibly more data.
There are no other restrictions on
the form of the login name, password, and timestamp.

If the password is unacceptable,
.B checkpassword
exits 1.
If
.B checkpassword
is misused,
it may instead exit 2.
If there is a temporary problem checking the password,
.B checkpassword
exits 111.

If the password is acceptable,
.B checkpassword
uses
.B execvp
to run
.B subprogram
with the given arguments.
.SH "CHECKPASSWORD-COMPATIBLE TOOLS"
There are other tools that offer the same interface as
.BR checkpassword .
Applications that use
.B checkpassword
are encouraged to take the
.B checkpassword
name as an argument,
so that they can be used with different
.BR checkpassword -compatible
tools.

Note that these tools do not follow the
.B getopt
interface.
Optional features are controlled through
(1) the tool name and
(2) environment variables.
.SH "THE PASSWORD DATABASE"
.B checkpassword
checks the login name and password against
.BR /etc/passwd ,
using the operating system's
.B getpwnam
and
.B crypt
functions.
It rejects accounts with empty passwords.
It ignores the timestamp.

Other
.BR checkpassword -compatible
tools have different interpretations of
login names, passwords, and timestamps.
Both the login name and the password
should be treated as secrets by the application calling
.BR checkpassword ;
the only distinction is for administrative convenience.
The timestamp should include any other information
that the password is based on;
for example, the challenge in a challenge-response system such as APOP.

.B WARNING:
.B getpwnam
is inherently unreliable.
It fails to distinguish between temporary errors and nonexistent users.
Future versions of
.B getpwnam
should return ETXTBSY to indicate temporary errors
and ESRCH to indicate nonexistent users.
.SH "PROCESS-STATE CHANGES"
Before invoking
.BR subprogram ,
.B checkpassword
sets up
.BR $USER ,
.BR $HOME ,
.BR $SHELL ,
its supplementary groups,
its gid,
its uid,
and its working directory.

Other
.BR checkpassword -compatible
tools may make different changes to the process state.
These effects must be documented;
applications will differ in their requirements.
.SH "VERSION"
This documentation describes
.B checkpassword
version 0.80.
See
.B http://pobox.com/~djb/checkpwd.html
for updates.
.SH "SEE ALSO"
getpwnam(3),
crypt(3)


debian/changelog
qmail (1.03-45~0njl.1) etch-backports; urgency=low

  * Backport to Etch
  * Add net-qmail-tls patch from
    http://inoa.net/qmail-tls/netqmail-1.05-tls-20070408.patch
  * Add qmail-1.03-fastremote-3.patch
  * Add sendmail-flagf.patch
  * Remove qmail-0.0.0.0.patch
  * Add qmail-1.03-moreipme-0.5.patch
  * NB, qmail-ldap already includes oversized-dns.patch and ext-todo patch.

 -- Nick Leverton <nick@leverton.org>  Fri, 16 May 2008 23:17:54 +0100

qmail (1.03-45) unstable; urgency=low

  * Updated prerm to remove dependancy on update-inetd (Closes #439702)

 -- Jon Marler <jmarler@debian.org>  Tue, 11 Dec 2007 15:32:58 -0600

qmail (1.03-44) unstable; urgency=low

  * One minor fix in the control file for autobuild compatibility

 -- Jon Marler <jmarler@debian.org>  Fri, 19 Jan 2007 07:53:39 -0600

qmail (1.03-43) unstable; urgency=low

  * Updated to Debian Policy Standards-Version 3.7.2
  * Fixed some non-essential errors in the control file
  * Removed old 1.02 transition code and usr/doc cleanup (Closes: #406990)

 -- Jon Marler <jmarler@debian.org>  Sun, 31 Dec 2006 02:47:14 -0600

qmail (1.03-42) unstable; urgency=low

  * Attempt #2 at fixing autobuild bug (Closes: #402331)

 -- Jon Marler <jmarler@debian.org>  Thu, 21 Dec 2006 16:54:58 -0600

qmail (1.03-41) unstable; urgency=low

  * Changed rules file to allow building -src on systems without qmail users
    (Closes: #402331)

 -- Jon Marler <jmarler@debian.org>  Sat, 16 Dec 2006 12:30:32 -0600

qmail (1.03-40) unstable; urgency=low

  * Disabled all debconf notes per Christian Perrier <bubulle@debian.org>
    Joey Hess <joeyh@debian.org>, and srivasta@debian.org (Closes: #388952)

 -- Jon Marler <jmarler@debian.org>  Wed, 27 Sep 2006 13:40:16 -0500

qmail (1.03-39) unstable; urgency=low

  * Added debconf-2.0 dependancy option

 -- Jon Marler <jmarler@debian.org>  Mon, 26 Sep 2005 13:29:40 -0500

qmail (1.03-38) unstable; urgency=high

  * Added ISO C patch taken from netqmil to fix 64-bit crashes (Closes: #309048)

 -- Jon Marler <jmarler@debian.org>  Sun, 15 May 2005 11:22:27 -0500

qmail (1.03-37) unstable; urgency=high

  * Updated qregex patch to 20040725 version
  * Added Build-Depends line (Closes: #302677)
  * Added pidfile settings to init file (Closes: #293484)
  * Applied 0.0.0.0 local patch (Closes: #226753)
  * Increased default ulimit to 16384 for x86_64 (Closes: #291437)
  * Applied localtime patch (Closes: #272072)
  * Fixed bug in qmail-procmail to allow arguments (Closes: #262906)
  * Updated postrm to remove /etc/qmail/users on purge (Closes: #272806)

 -- Jon Marler <jmarler@debian.org>  Thu,  5 May 2005 11:54:04 -0600

qmail (1.03-36) unstable; urgency=high

  * Fixed problem when building with apt-get source -b (Closes: #280677)

 -- Jon Marler <jmarler@debian.org>  Thu, 11 Nov 2004 13:03:39 -0600

qmail (1.03-35) unstable; urgency=high

  * Fixed debconf errors (Closes: #272248, #272164, #272797, #272804)
  * Added testing to release list

 -- Jon Marler <jmarler@debian.org>  Thu, 28 Oct 2004 11:15:19 -0500

qmail (1.03-34) unstable; urgency=low

  * Moved logcheck ignore file to docs directory (Closes: #271118)
  * File is already included in logcheck-database which seems more logical

 -- Jon Marler <jmarler@debian.org>  Mon, 13 Sep 2004 06:50:39 -0500

qmail (1.03-33) unstable; urgency=low

  * Added -R -H to example POP3 server call (Closes: #269119)
  * Added sample logcheck ignore files (Closes: #266143)
  * Fixed hang bug in postinst (Closes: #266010)

 -- Jon Marler <jmarler@debian.org>  Tue, 31 Aug 2004 16:36:51 -0500

qmail (1.03-32) unstable; urgency=low

  * Initial debconf implementation
    - Please file bug reports for questions i missed
  * Removed reference to bmtpatch. (Closes: #246574)
  * Leaving bmtpatch in -src pacakge as many users still use it
  * Applied latest qregex patch (Closes: #246408)

 -- Jon Marler <jmarler@debian.org>  Tue, 13 Jul 2004 10:24:49 -0500

qmail (1.03-31) unstable; urgency=low

  * Added dependancy on perl-modules (Closes: #233015)

 -- Jon Marler <jmarler@debian.org>  Tue, 17 Feb 2004 13:03:08 -0600

qmail (1.03-30) unstable; urgency=high

  * Added patch from Scott Gifford to correct qmail-smtpd crash

 -- Jon Marler <jmarler@debian.org>  Tue, 20 Jan 2004 11:57:18 -0600

qmail (1.03-29) unstable; urgency=low

  * Added patch to resolve new glibc errors (Closes: #218564, #219872)
  * Added patch to fix errno declaration errors (Closes: #219872)
  * Remove strict barewords from preinst (Closes: #215152)
  * Fixed minor type in build-qmail (Closes: #213415)
  * Added patch to make qmail-local behave better (Closes: #95353)

 -- Jon Marler <jmarler@debian.org>  Thu, 13 Nov 2003 17:13:33 -0600

qmail (1.03-28) unstable; urgency=low

  * Fixed logging in init to handle rblsmtpd (Closes: #192116)
  * Added ability to override qmail installation check (Closes: #187452)
  * Changed groff dependancy to groff-base (Closes: #181021)
  * Removed rm -rf /etc/qmail command in postrm (Closes: #179637)
  * Added list of pre-applied patches in build script before binary build

 -- Jon Marler <jmarler@debian.org>  Thu, 25 Sep 2003 16:48:34 -0500

qmail (1.03-27) unstable; urgency=low

  * Added stat to usage list (Closes: #179831)

 -- Jon Marler <jmarler@debian.org>  Tue,  4 Feb 2003 16:16:41 -0600

qmail (1.03-26) unstable; urgency=low

  * Added -isp flag to dpkg-gencontrol (Closes: #179609)
  * Removed emacs settings from changelog

 -- Jon Marler <jmarler@debian.org>  Mon,  3 Feb 2003 09:56:19 -0600

qmail (1.03-25) unstable; urgency=low

  * Applied patch from thunder7@xs4all.nl for libc6 (Closes: #174408)
  * Fixed typo in postinst (Closes: #134011)
  * Added warning in build script regarding tmpfs (Closes: #155024)
  * Removed extra -f from postinst (Closes: #155573)
  * Updated init.d script (Closes: #128638)

 -- Jon Marler <jmarler@debian.org>  Thu, 20 Jan 2003 02:56:56 -0600

qmail (1.03-24) testing unstable; urgency=low


contrib/checkpassword-0.90/checkpassword.c
#include "error.h"
#include "pathexec.h"
#include "prot.h"

extern char *crypt();
#include <pwd.h>
static struct passwd *pw;

#include "hasspnam.h"
#ifdef HASGETSPNAM
#include <shadow.h>
static struct spwd *spw;
#endif

#include "hasuserpw.h"
#ifdef HASUSERPW
#include <userpw.h>
static struct userpw *upw;
#endif

static char up[513];
static int uplen;

main(int argc,char **argv)
{
  char *login;
  char *password;
  char *encrypted;
  char *stored;
  int r;
  int i;
 
  if (!argv[1]) _exit(2);
 
  uplen = 0;
  for (;;) {
    do
      r = read(3,up + uplen,sizeof(up) - uplen);
    while ((r == -1) && (errno == error_intr));
    if (r == -1) _exit(111);
    if (r == 0) break;
    uplen += r;
    if (uplen >= sizeof(up)) _exit(1);
  }
  close(3);

  i = 0;
  if (i >= uplen) _exit(2);
  login = up + i;
  while (up[i++]) if (i >= uplen) _exit(2);
  password = up + i;
  if (i >= uplen) _exit(2);
  while (up[i++]) if (i >= uplen) _exit(2);

  pw = getpwnam(login);
  if (pw)
    stored = pw->pw_passwd;
  else {
    if (errno == error_txtbsy) _exit(111);
    _exit(1);
  }
#ifdef HASUSERPW
  upw = getuserpw(login);
  if (upw)
    stored = upw->upw_passwd;
  else
    if (errno == error_txtbsy) _exit(111);
#endif
#ifdef HASGETSPNAM
  spw = getspnam(login);
  if (spw)
    stored = spw->sp_pwdp;
  else
    if (errno == error_txtbsy) _exit(111);
#endif
  if (!stored) _exit(1);
 
  encrypted = crypt(password,stored);
  for (i = 0;i < sizeof(up);++i) up[i] = 0;
 
  if (!*stored || strcmp(encrypted,stored)) _exit(1);
 
  if (prot_gid((int) pw->pw_gid) == -1) _exit(1);
  if (prot_uid((int) pw->pw_uid) == -1) _exit(1);
  if (chdir(pw->pw_dir) == -1) _exit(111);

  if (!pathexec_env("USER",pw->pw_name)) _exit(111);
  if (!pathexec_env("HOME",pw->pw_dir)) _exit(111);
  if (!pathexec_env("SHELL",pw->pw_shell)) _exit(111);
  pathexec(argv + 1);
  _exit(111);
}
